src/macros.rs

```diff
@@ -33,38 +33,209 @@ macro_rules! scopes {
-            .insert($p1.to_string(), json!($p2))
+macro_rules! opt {
+    (, $default:expr) => {
+        $default
     };
+    ($optional:expr, $default:expr) => {
+        $optional
+    };
+}
+
+/// Count items in a list of items within a macro, taken from here:
+/// https://danielkeep.github.io/tlborm/book/blk-counting.html
+#[doc(hidden)]
+#[macro_export]
+macro_rules! replace_expr {
+    ($_t:tt $sub:expr) => {
+        $sub
+    };
+}
+#[doc(hidden)]
+#[macro_export]
+macro_rules! count_items {
+    ($($item:expr),*) => {<[()]>::len(&[$($crate::replace_expr!($item ())),*])};
+}
+
+/// Hack to convert an identifier to a string, including raw identifiers.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! ident_str {
+    ($name:ident) => {{
+        const NAME: &str = stringify!($name);
+        match &NAME.get(..2) {
+            Some("r#") => &NAME[2..],
+            _ => NAME,
+        }
+    }};
+}
+
+/// Private macro to insert either required or optional fields. Pattern matching
+/// will accordingly pick the branch, and then insert ($key, $val) into $map.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! params_internal {
+    ($map:ident, req, $name:ident, $key:expr, $val:expr) => {
+        $map.insert($key, $val);
+    };
+    ($map:ident, opt, $name:ident, $key:expr, $val:expr) => {
+        // Will only insert when `$name` is not None.
+        if let Some(ref $name) = $name {
+            $map.insert($key, $val);
+        }
+    };
+}
+
+/// This macro and [`build_json`] help make the endpoints as concise as possible
+/// and boilerplate-free, which is specially important when initializing the
+/// parameters of the query with a HashMap/similar. Their items follow the
+/// syntax:
+///
+///   [req|opt] key [=> value]
+///
+/// The first keyword is just to distinguish between a direct insert into the
+/// hashmap (required parameter), and an insert only if the value is `Some(...)`
+/// (optional parameter). This is followed by the variable to be inserted, which
+/// shall have the same name as the key in the map (meaning that `r#type` may
+/// have to be used).
+///
+/// It also works similarly to how struct initialization works. You may provide
+/// a key and a value with `MyStruct { key: value }`, or if both have the same
+/// name as the key, `MyStruct { key }` is enough.
+///
+/// For more information, please refer to the `test::test_build_map` function in
+/// this module to see an example, or the real usages in Rspotify's client.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! build_map {
+    (
+        $(
+            $kind:ident $name:ident $( => $val:expr )?
+        ),* $(,)?
+    ) => {{
+        let mut params = $crate::http::Query::with_capacity(
+            $crate::count_items!($( $name ),*)
+        );
+        $(
+            $crate::params_internal!(
+                params,
+                $kind,
+                $name,
+                $crate::ident_str!($name),
+                $crate::opt!($( $val )?, $name)
```

Comment:
Ooooh, this macro is a little complex, I take times to understand it, and the key point is `$( => $val:expr )?`, the `?` means "at most one" repetition.

if `$( => $val:expr )` doesn't repeat, then the `$crate::opt!($( $val )?, $name)` will match `(, $default:expr)` arm, set the `$name` as `$val`

otherwise it will match `($optional:expr, $default:expr)`, set the `$( $val )` as `$val`.

Reply:
Yup, that's it. The thing is that if ` $( => $val:expr )?` doesn't repeat is output is empty, and I needed a "default" value.